eclair-core/src/main/scala/fr/acinq/eclair/tor/Socks5Connection.scala

```diff
@@ -0,0 +1,169 @@
+      context become greetings(sender(), c)
+      underlying ! Tcp.Register(self)
+      underlying ! Tcp.ResumeReading
+      log.info(s"send greeting")
+      underlying ! Tcp.Write(socks5Greeting)
+  }
+
+  def greetings(commander: ActorRef, connectCommand: Socks5Connect): Receive = {
+    case Tcp.Received(data) =>
+      log.info(s"receive auth")
+      try {
+        if (data(0) != 0x05) {
+          throw new RuntimeException("Invalid SOCKS5 proxy response")
+        } else if (data(1) != 0x00) {
+          throw new RuntimeException("Unrecognized SOCKS5 auth method")
+        } else {
+          context become connectionRequest(commander, connectCommand)
+          log.info(s"send con req")
+          underlying ! Tcp.Write(socks5ConnectionRequest(connectCommand.address))
+          underlying ! Tcp.ResumeReading
+        }
+      } catch {
+        case e: Throwable => handleErrors("Error connecting to SOCKS5 proxy", commander, connectCommand, e)
+      }
+    case c: Tcp.ConnectionClosed =>
+      commander ! c
+      context stop self
+  }
+
+  def connectionRequest(commander: ActorRef, connectCommand: Socks5Connect): Receive = {
+    case c@Tcp.Received(data) =>
+      log.info(s"receive address")
+      try {
+      if (data(0) != 0x05) {
+        throw new RuntimeException("Invalid SOCKS5 proxy response")
+      } else {
+          val status = data(1)
+          if (status != 0) {
+            throw new RuntimeException(connectErrors.getOrElse(status, s"Unknown SOCKS5 error $status"))
+          }
+          val connectedAddress = data(3) match {
+            case 0x01 =>
+              val ip = Array(data(4), data(5), data(6), data(7))
+              val port = data(8).toInt << 8 | data(9)
+              new InetSocketAddress(InetAddress.getByAddress(ip), port)
+            case 0x03 =>
+              val len = data(4)
+              val start = 5
+              val end = start + len
+              val domain = data.slice(start, end).utf8String
+              val port = data(end).toInt << 8 | data(end + 1)
+              new InetSocketAddress(domain, port)
+            case 0x04 =>
+              val ip = Array.ofDim[Byte](16)
+              data.copyToArray(ip, 4, 4 + ip.length)
+              val port = data(4 + ip.length).toInt << 8 | data(4 + ip.length + 1)
+              new InetSocketAddress(InetAddress.getByAddress(ip), port)
+            case _ => throw new RuntimeException(s"Unrecognized address type")
+          }
+          context become connected
+          log.info(s"connected $connectedAddress")
+          commander ! Socks5Connected(connectedAddress)
+      }
+      } catch {
+        case e: Throwable => handleErrors("Cannot establish SOCKS5 connection", commander, connectCommand, e)
+      }
+    case c: Tcp.ConnectionClosed =>
+      commander ! c
+      context stop self
+  }
+
+  def connected: Receive = {
+    case Tcp.Register(actor, keepOpenOnPeerClosed, useResumeWriting) =>
+      handler = actor
```

Comment:
This is most certainly where your problem lies. You should forward the `Tcp.Register` message to the actual TCP connection, so that it knows that incoming messages should be sent to the `TransportHandler`, not to the `Socks5Connection`.